src/charm.py

```diff
@@ -19,6 +24,181 @@ class KratosCharm(CharmBase):
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        pebble_layer = {
+            "summary": "kratos layer",
+            "description": "pebble config layer for kratos",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Kratos Operator layer",
+                    "startup": "enabled",
+                    "command": "kratos serve all --config /etc/config/kratos.yaml",
+                }
+            },
+            "checks": {
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4434/admin/health/ready"},
+                },
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4434/admin/health/alive"},
+                },
+            },
+        }
+        return Layer(pebble_layer)
+
+    @property
+    def _config(self) -> str:
+        db_info = self._get_database_relation_info()
+        config = {
+            "log": {"level": "trace"},
+            "identity": {
+                "default_schema_id": "default",
+                "schemas": [{"id": "default", "url": f"file://{self._identity_schema_file_path}"}],
+            },
+            "selfservice": {
+                "default_browser_return_url": "http://127.0.0.1:9999/",
+                "flows": {
+                    "registration": {
+                        "enabled": True,
+                        "ui_url": "http://127.0.0.1:9999/registration",
+                    }
+                },
+            },
+            "dsn": f"postgres://{db_info['username']}:{db_info['password']}@{db_info['endpoints']}/postgres",
+            "courier": {
+                "smtp": {
+                    # TODO: dynamic connection uri through charm config
+                    "connection_uri": "smtps://test:test@mailslurper:1025/?skip_ssl_verify=true"
+                }
+            },
+        }
+
+        return yaml.dump(config)
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer and kratos config if changed."""
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._pebble_layer
+
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            self._container.replan()
+
+        # Compare changes in kratos config
+        try:
+            current_config = self._container.pull(self._config_file_path).read()
+        except PathError:
+            self._container.push(self._config_file_path, self._config, make_dirs=True)
```

Comment:
To avoid the charm getting into error state when the directory doesn't exist yet, we could push the config file while comparing the layers.
They will differ the first time they're compared, and we can assume that the config will not exist neither. We could force-push it once and then compare against the current state.